src/Analysis/Ast/Impl/Analyzer/ProgressReporter.cs

```diff
@@ -0,0 +1,95 @@
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Threading;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.Services;
+
+namespace Microsoft.Python.Analysis.Analyzer {
+    internal sealed class ProgressReporter : IProgressReporter, IDisposable {
+        private const int _initialDelay = 100;
+        private const int _reportingInterval = 200;
+        private const int _disposeInterval = 2000;
+
+        private readonly IProgressService _progressService;
+        private readonly object _lock = new object();
+        private int _lastReportedCount;
+        private int _lastReportedToClient;
+        private IProgress _progress;
+        private Timer _reportTimer;
+        private Timer _disposeTimer;
+        private bool _running;
+        private int _disposedId;
+
+        public ProgressReporter(IProgressService progressService) {
+            _progressService = progressService;
+        }
+
+        public void Dispose() {
+            lock (_lock) {
+                EndReport();
+            }
+        }
+
+        public void ReportRemaining(int count = -1) {
+            lock (_lock) {
+                if (!_running) {
+                    // Delay reporting a bit in case the analysis is short in order to reduce UI flicker.
+                    _running = true;
+                    _reportTimer?.Dispose();
+                    _reportTimer = new Timer(OnReportTimer, null, _initialDelay, _reportingInterval);
+                }
+
+                _disposeTimer?.Dispose();
+                _disposeTimer = new Timer(OnDisposeTimer, ++_disposedId, _disposeInterval, Timeout.Infinite);
+                if (count >= 0) {
+                    _lastReportedCount = count;
+                }
+            }
+        }
+
+        private void OnReportTimer(object o) {
+            lock (_lock) {
+                if (_running && _lastReportedToClient != _lastReportedCount) {
+                    _lastReportedToClient = _lastReportedCount;
+                    _progress = _progress ?? _progressService.BeginProgress();
+                    if (_lastReportedCount > 0) {
+                        _progress?.Report(Resources.AnalysisProgress.FormatUI(_lastReportedCount)).DoNotWait();
+                    }
+                }
+            }
+        }
+
+        private void OnDisposeTimer(object o) {
```

Comment:
For me, if the number doesn't change for a second or so, the message disappears, and then never returns again even if the analysis continues. I think it's because this function gets called and everything gets disposed early.

Reply:
So you have Microsoft/vscode-python#4620? You'd need to get new code and F5 from the extension.

Reply:
Also, the message does can get removed if update comes later than the dispose interval. This is normal at least for now.

Reply:
It's pretty evident when you're working on a big repo like scikit-learn (`k_means_.py`, as usual).

Reply:
There is no known way to learn that there is more to come. Not yet anyway.

Reply:
I don't understand that statement; If the LS reports 500 things left to analyze, then pauses for 5 seconds, the text disappears. Then it sees that there are say 480 things left to analyze, but that message doesn't seem to be emitted. I'm not saying that it should know the future, but I wouldn't expect that the timeout would prevent all future indications to be thrown away or never rerun.

Reply:
It does not dispose the report forever, it hides the progress. Next message will re-create the progress.

Reply:
That's what I'm saying doesn't always happen. Screencap:

![687](https://user-images.githubusercontent.com/5341706/53830259-e12b3580-3f36-11e9-9ad2-26d7200ea552.gif)

Note how it disappears sometimes, then disappears completely.

Though, modifying this file (which is imported in many other places) does cause it to show briefly. Maybe it's just this initial analysis that does it, I just thought it was strange for it to be gone altogether when there are so many things left to analyze and it's logging progress.

Reply:
If it reports *complete*, the analysis is done and there is nothing to report.

Reply:
I understand, but I can see real CPU usage (even without logging) after the message has disappeared. I don't believe the analysis is complete when the message goes away at 1185 items remaining in my screencap above.

Reply:
Point is - there is no reliable progress reporting in analysis graph at the moment. Showing nothing is worse than at least giving some indication of work getting done. Not disposing of the message does not necessarily mean it will be showing something meaningful, most probably 0 or stale static data until the next update comes. There is an event when analysis complete fully, it may be possible to wire it up - but it wont be perfect still.

Reply:
I can see from adding some extra logging that ReportRemaining stops getting called at some point even though there is some analysis being done. I guess this doesn't have to do with disposing at all (which I expected), but just generally how this first implementation writes the number remaining, which could be improved in the future.